From ef17a32e41a2379707ff23f45625b2dd5825e7bb Mon Sep 17 00:00:00 2001
From: William Douglas <william.r.douglas@gmail.com>
Date: Thu, 26 May 2022 17:46:59 +0000
Subject: [PATCH] Add Clear Linux kvm to native switch

With the kvm kernel being removed and the native kernel being intended
for use in its place, add logic to detect the kvm kernel is the
default type and switch it to native. Also don't keep the kvm kernel
tip around after the native kernel has been used.

Signed-off-by: William Douglas <william.r.douglas@gmail.com>
---
 src/bootman/update.c | 10 +++++++++-
 1 file changed, 9 insertions(+), 1 deletion(-)

diff --git a/src/bootman/update.c b/src/bootman/update.c
index b9c06bb..a445b50 100644
--- a/src/bootman/update.c
+++ b/src/bootman/update.c
@@ -287,7 +287,7 @@ static bool boot_manager_update_native(BootManager *self)
                                 }
                                 LOG_INFO("update_native: not-running: %s", tk->source.path);
                                 /* Preserve tip */
-                                if (tip && tk == tip) {
+                                if (tip && tk == tip && strcmp(kernel_type, "kvm") != 0) {
                                         LOG_DEBUG("update_native: Skipping default-%s: %s",
                                                   kernel_type,
                                                   tk->source.path);
@@ -323,9 +323,17 @@ static bool boot_manager_update_native(BootManager *self)
                 if (system_kernel && system_kernel->ktype[0] != '\0') {
                         new_default =
                             boot_manager_get_default_for_type(self, kernels, system_kernel->ktype);
+                        if (new_default && strcmp(new_default->meta.ktype, "kvm") == 0) {
+                                new_default =
+                                        boot_manager_get_default_for_type(self, kernels, "native");
+                        }
                 }
         } else {
                 new_default = boot_manager_get_default_for_type(self, kernels, running->meta.ktype);
+                if (new_default && strcmp(new_default->meta.ktype, "kvm") == 0) {
+                        new_default =
+                                boot_manager_get_default_for_type(self, kernels, "native");
+                }
         }
 
         if (new_default) {
-- 
2.36.0

